### DISCUSSION TITLE
Hinode - problem

### DISCUSSION THREAD

**Discussion by maciejglowackimg** [gethinode/hinode#763] [Help]
Hi @markdumay, i have a little problem. I try add to hinode "https://myrthos.net/docs/shortcodes/mgallery/installation/". (what is the correct addition procedure?) I use config.toml and go.mod files but in instruction is config/_default/params/params.toml.

I managed to install the changes, I rolled them back and I got this message...

ERROR render of "page" failed: execute of template failed: template: partials/head/head.html:20:7: executing "head" at <partial "head/seo.html" .>: error calling partial: "/Users/witchcraft/Library/Caches/hugo_cache/modules/filecache/modules/pkg/mod/github.com/gethinode/hinode@v0.22.5/layouts/partials/head/seo.html:10:23": execute of template failed: template: partials/head/seo.html:10:23: executing "partials/head/seo.html" at <partial "utilities/GetDescription.html" .>: error calling partial: "/Users/witchcraft/Library/Caches/hugo_cache/modules/filecache/modules/pkg/mod/github.com/gethinode/mod-utils/v2@v2.0.3/layouts/partials/utilities/GetDescription.html:1:13": execute of template failed: template: partials/utilities/GetDescription.html:1:13: executing "partials/utilities/GetDescription.html" at <.page>: can't evaluate field page in type page.Page

**Comment by markdumay** [comment]
I suspect myrthos is using a forked version of hinode that is still in beta. The beta introduces a [breaking change](https://github.com/gethinode/hinode/releases/tag/v0.23.0-alpha8). You want to make sure your hinode version is compatible with the one myrthos is using. You can do so by adjusting go.mod as needed.

**Reply by maciejglowackimg** [reply]
@markdumay  0.22.5 is latest relase, is correct?
<img width="713" alt="Zrzut ekranu 2024-02-9 o 20 57 16" src="https://github.com/gethinode/hinode/assets/131944141/1a93e5ae-414d-4cb9-b47f-1265d4b9d97f">

**Reply by maciejglowackimg** [reply]
@markdumay  0.20.6 work ... 
<img width="658" alt="Zrzut ekranu 2024-02-9 o 22 17 36" src="https://github.com/gethinode/hinode/assets/131944141/c44b52b3-e71a-4493-98fe-7f909af7890c">

**Reply by markdumay** [reply]
[v0.22.5](https://github.com/gethinode/hinode/releases/tag/v0.22.5) is the latest stable version indeed. Does v0.20.6 work for you now?

**Reply by maciejglowackimg** [reply]
@markdumay v0.22.5 don't work but v0.20.6 work correctly - I don't know why, maybe some higher ones will also work, but 0.22.5 does not work, there is an error as in my first entry.

**Comment by dims31** [comment]
Hello,

Same here after updating mod-utils...

[Exemple](https://github.com/dims31/tdm/tree/update)

I think it concern this [change](https://github.com/gethinode/mod-utils/releases/tag/v2.0.0) but I can't figure to work with this... 

ERROR:
ERROR render of "page" failed: execute of template failed: template: partials/head/head.html:20:7: executing "head" at <partial "head/seo.html" .>: error calling partial: "C:\Users\dimit\Desktop\git\tdm\_vendor\github.com\gethinode\hinode\layouts\partials\head\seo.html:10:23": execute of template failed: template: partials/head/seo.html:10:23: executing "partials/head/seo.html" at <partial "utilities/GetDescription.html" .>: error calling partial: "C:\Users\dimit\Desktop\git\tdm\_vendor\github.com\gethinode\mod-utils\v2\layouts\partials\utilities\GetDescription.html:1:13": execute of template failed: template: partials/utilities/GetDescription.html:1:13: executing "partials/utilities/GetDescription.html" at <.page>: can't evaluate field page in type page.Page
Built in 3019 ms
Error: error building site: render: failed to render pages: render of "page" failed: execute of template failed: template: partials/head/head.html:20:7: executing "head" at <partial "head/seo.html" .>: error calling partial: "C:\Users\dimit\Desktop\git\tdm\_vendor\github.com\gethinode\hinode\layouts\partials\head\seo.html:10:23": execute of template failed: template: partials/head/seo.html:10:23: executing "partials/head/seo.html" at <partial "utilities/GetDescription.html" .>: error calling partial: "C:\Users\dimit\Desktop\git\tdm\_vendor\github.com\gethinode\mod-utils\v2\layouts\partials\utilities\GetDescription.html:1:13": execute of template failed: template: partials/utilities/GetDescription.html:1:13: executing "partials/utilities/GetDescription.html" at <.page>: can't evaluate field page in type page.Page

**Reply by markdumay** [reply]
It is related indeed. The module `mod-utils` was updated to a major new [release v2](https://github.com/gethinode/mod-utils/releases/tag/v2.0.0) as it introduces the breaking change as mentioned. I follow the [semantic versioning](https://semver.org) pattern to manage the versions and releases.

Unless you want to switch to the Hinode beta (which is not stable yet), you should revert back to the `v1` release of `mod-utils`. Modify the file `go.mod` in your repository root as needed. For example, the same [file](https://github.com/gethinode/hinode/blob/e7a65822297ad8ed41179c24ded11325a6165590/go.mod) in Hinode v0.22.5 (the latest stable release) uses the following configuration:

```go
module github.com/gethinode/hinode

go 1.19

require (
	github.com/airbnb/lottie-web v5.12.2+incompatible // indirect
	github.com/gethinode/mod-bootstrap v1.2.2 // indirect
	github.com/gethinode/mod-flexsearch v1.8.1 // indirect
	github.com/gethinode/mod-fontawesome v1.8.0 // indirect
	github.com/gethinode/mod-katex v1.0.5 // indirect
	github.com/gethinode/mod-leaflet v0.4.0 // indirect
	github.com/gethinode/mod-lottie v1.4.1 // indirect
	github.com/gethinode/mod-utils v1.0.2 // indirect
	github.com/nextapps-de/flexsearch v0.0.0-20240108021025-afd75f742f22 // indirect
	github.com/twbs/bootstrap v5.3.2+incompatible // indirect
)
```

For comparison, the current beta uses the following configuration:

```go
module github.com/gethinode/hinode

go 1.19

require (
	github.com/airbnb/lottie-web v5.12.2+incompatible // indirect
	github.com/gethinode/mod-bootstrap v1.2.2 // indirect
	github.com/gethinode/mod-flexsearch v1.9.0 // indirect
	github.com/gethinode/mod-fontawesome v1.8.1 // indirect
	github.com/gethinode/mod-katex v1.0.5 // indirect
	github.com/gethinode/mod-leaflet v1.0.0 // indirect
	github.com/gethinode/mod-lottie v1.4.2 // indirect
	github.com/gethinode/mod-utils/v2 v2.0.3 // indirect
	github.com/nextapps-de/flexsearch v0.0.0-20240110101704-4c3966709f85 // indirect
	github.com/twbs/bootstrap v5.3.2+incompatible // indirect
)
```

Observe the `github.com/gethinode/mod-utils/v2` versus `github.com/gethinode/mod-utils` declaration. Go (and Hugo) requires major releases to be added as suffix to the import path. Accordingly, Hugo will never upgrade to a new major release automatically. 

If you use `npm`, you can run `npm run mod:update` to update the latest version of Hinode and required modules. The updated modules should still use the latest compatible release.

If needed, you can manually update your own codebase. Take a look at the files that reference `GetDesciption` with the new [calling convention](https://github.com/search?q=repo%3Agethinode%2Fhinode%20getdescription&type=code).

Let me know if it still fails. Keeping versions in sync can be a crime sometimes.

**Reply by maciejglowackimg** [reply]
@markdumay with your configuration go.mod work fine, the problem may be that the "hugo mod get -u" function doesn't work properly.

**Reply by dims31** [reply]
@markdumay, I manage update with npm, and modules were updated to new major release without action on my side...

I revert back to old go.mod and it works perfectly but if I update hugo module, it takes the new version automatically (with mod-utils v2)

My go.mod file after run update with npm :

`
require (
	github.com/airbnb/lottie-web v5.12.2+incompatible // indirect
	github.com/gethinode/hinode v0.22.5 // indirect
	github.com/gethinode/mod-bootstrap v1.2.2 // indirect
	github.com/gethinode/mod-flexsearch v1.9.0 // indirect
	github.com/gethinode/mod-fontawesome v1.8.1 // indirect
	github.com/gethinode/mod-katex v1.0.5 // indirect
	github.com/gethinode/mod-leaflet v1.0.0 // indirect
	github.com/gethinode/mod-lottie v1.4.2 // indirect
	github.com/gethinode/mod-utils v1.1.0 // indirect
	github.com/gethinode/mod-utils/v2 v2.0.3 // indirect
	github.com/nextapps-de/flexsearch v0.0.0-20240110101704-4c3966709f85 // indirect
	github.com/twbs/bootstrap v5.3.2+incompatible // indirect
)
`

Thank you for your help

**Reply by markdumay** [reply]
Hugo modules work slightly different than most package managers, such as npm. Any call to `hugo mod get` will not only download the modules, but also **update them to the latest version**. See this [thread](https://discourse.gohugo.io/t/strange-module-issue/47796/2) for more context. This is far from ideal, because preferably you want to use version pinning during builds and only use newer versions using an explicit instruction.

If I remember correctly, a call to `hugo mod vendor` will not update the modules but download the current versions only. When running `hugo` or `hugo server` subsequently, hugo will pick up the vendored modules instead. You may need to revise your npm scripts, e.g. removing any calls that include `hugo mod get` (such as `mod:update`). Be sure to address any `pre` scripts too (e.g. `prestart` and `prebuild`), as npm automatically invokes them prior to a regular command.

Alternatively, you could revise the code that is incompatible with `mod-utils/v2` and be done with it. It is a relatively simple adjustment, see below:

```go
// old partial invocation
partial "utilities/GetDescription.html" .

// new partial invocation
partial "utilities/GetDescription.html" (dict "page" .)
```

**Reply by dims31** [reply]
Thank you for the clarification,

If I understand, when updating, Hugo doesn't track version of submodules for a specific hinode version, and instead of download all latest version of submodules. As you said It is not ideal...

I will not update Hinode until new version is released.

Have a good day!

**Reply by markdumay** [reply]
Sure! Just a few clarifications:
- hugo does respect module versions, however, `hugo mod get` downloads **and upgrades** modules. You should be safe just running `hugo` or `hugo server`. You do need to modify the `package.json` file though, as Hinode includes the upgrade by default.
- Module vendoring is required for the CSS purge to work properly, which is a Hinode-specific requirement. I experienced some edge cases, so by default I always vendor my modules.
- It would be better to refer to `modules` instead of `submodules` to avoid confusion with [git submodules](https://git-scm.com/book/en/v2/Git-Tools-Submodules). In the past, Hugo did not support Go modules yet, and relied on git submodules for themes to work.
